lib/scopes/cache/cache.go

```diff
@@ -0,0 +1,272 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package cache
+
+import (
+	"iter"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/scopes"
+)
+
+// Config configures a cache.
+type Config[T any, K comparable] struct {
+	// Scope is the function used to determine the scope of a value.
+	Scope func(T) string
+	// Key is the function used to determine the primary key of a value.
+	Key func(T) K
+}
+
+// node is a tree node in the cache which stores values at a given scope.
+type node[T any, K comparable] struct {
+	// members is the set of values "at" this scope.
+	members map[K]struct{}
+
+	// children is the set of child scopes.
+	children map[string]*node[T, K]
+}
+
+// Cache is a generic scoped value cache. It constructs a basic tree structure based on scope segments
+// to support efficient lookings up values based on common scoping patterns. This cache is intended to be
+// a tool for building more purpose-specific wrappers and should not be used directly in access-control
+// logic due to ease of misuse.
+//
+// NOTE: this is an early stage prototype and has a few notable limitations:
+//   - Not currently safe for concurrent use.
+//   - Iteration order of read methods is nondeterministic.
+//   - No cleanup of empty scopes.
+type Cache[T any, K comparable] struct {
+	cfg   Config[T, K]
+	items map[K]T
+	root  *node[T, K]
```

Comment:
I sort of suspect that a btreemap of scopes to set of keys would end up performing better while also being somewhat simpler to operate (depending on whether we end up needing `ResourcesSubjectToPolicyScope` more than `PoliciesApplicableToResourceScope`, with the former accessing a contiguous range while the latter needs `O(log N)` lookups for each scope component). WDYT? Something to look into if and when we have time, or something you have already considered and discarded?

Reply:
I anticipate `PoliciesApplicableToResourceScope` to be the more perf-critical.  `ResourcesSubjectToPolicyScope` will be the kind of lookup we do when deciding what resources a user can read when listing or whatever, so very few will be happening concurrently.

`PoliciesApplicableToResourceScope` is the kind of lookup that every agent will need to do on cache init in order to cache the set of scoped roles that might apply to that agent.  So likely high concurrency.

Hard to say wether or not we'll really need this kind of optimized lookup any time soon, but I wanted to make sure I had a good handle on how it would work _if_ we needed it, and at least have a basic untuned cache impl with the right "shape" as a starting point if we needed to do more focused perf tuning later.